[pt][aten] Enable the intra-op parallelism for layer norm

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29154 [bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256 (2/2)
* #29104 [bert/RoBERTa] Optimize LayerNorm with explicit vectorization using Vec256
* **#28464 [pt][aten] Enable the intra-op parallelism for layer norm**

We would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Before this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        14.10%           449.257ms        14.10%           449.257ms        9.360ms          NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```


After this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        1.55%            42.453ms         1.55%            42.453ms         884.448us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]     
```

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

jianyuh committed:
[pt][aten] Enable the intra-op parallelism for layer norm

We would like to enable the intra-op parallelism for layer norm.

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

[ghstack-poisoned]

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -34,28 +34,34 @@ void LayerNormKernelImplInternal(
-      Y_ptr[j] = (X_ptr[j] * scale + bias) * gamma_v + beta_v;
+  int num_tasks = at::get_num_threads();
+  at::parallel_for(0, num_tasks, 1, [&](int64_t start, int64_t end) {
+    int64_t M_per_thread = std::ceil(static_cast<float>(M) / num_tasks);
```
```suggestion
    const int64_t M_per_thread = (M + num_tasks - 1) / num_tasks;
```

jianyuh replied:
Updated. Thanks!

BIT-silence commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -34,28 +34,34 @@ void LayerNormKernelImplInternal(
+      T mean_val = T(0);
+      T rstd_val = T(0);
+      for (int64_t j = 0; j < N; ++j) {
+        mean_val += X_ptr[j];
```
Here we can consider to use vec256::Vec256<T> to do reduce. I remember in my local test, the compiler's auto vectorization is not as efficient as vec256 in reduce case. You can take a look at https://github.com/pytorch/pytorch/pull/23349

jianyuh replied:
Thanks! Do you know why  #23349 is not not merged into the PyTorch master branch?

BIT-silence replied:
No specific reasons. Just it is not a very urgent thing. So here since we will change this part, we can do it together.

jianyuh replied:
@BIT-silence : Added your PR in https://github.com/pytorch/pytorch/pull/29104.

jianyuh committed:
Update on "[pt][aten] Enable the intra-op parallelism for layer norm"


We would like to enable the intra-op parallelism for layer norm.

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

[ghstack-poisoned]

ilia-cher commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -34,28 +34,34 @@ void LayerNormKernelImplInternal(
+  at::parallel_for(0, num_tasks, 1, [&](int64_t start, int64_t end) {
+    const int64_t M_per_thread = (M + num_tasks - 1) / num_tasks;
+    const int64_t M_start = std::min(start * M_per_thread, M);
+    const int64_t M_end = std::min(end * M_per_thread, M);
```
this is not how `at::parallel_for` is typically used, you don't need to compute the number of tasks and start and end by yourself - check https://github.com/pytorch/pytorch/pull/19105 for example,i

you typically just need to write, e.g.
```
at::parallel_for(0, M, <choose grain size>, [&](int64_t start, int64_t end) { 
for (int64_t i = start; i < end; ++i) { ... }
});
```

jianyuh replied:
Thanks! Updated the PR. One thing to double check: When I set the `<choose grain size>` to 1, does this mean the loop is iterated in an interleaving way, or in a block-cyclic way?

jianyuh replied:
As @ilia-cher told me, the loop is iterated in a "block-cyclic" way: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L20

jianyuh committed:
Update on "[pt][aten] Enable the intra-op parallelism for layer norm"


We would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Before this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        14.10%           449.257ms        14.10%           449.257ms        9.360ms          NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```

https://docs.google.com/spreadsheets/d/137BkyMmVuLS0Shz7QHze-12CM1MEwhGimfIhh3P5kBY/edit#gid=375380546


After this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        1.55%            42.453ms         1.55%            42.453ms         884.448us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]     
```

https://docs.google.com/spreadsheets/d/137BkyMmVuLS0Shz7QHze-12CM1MEwhGimfIhh3P5kBY/edit#gid=930034990

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][aten] Enable the intra-op parallelism for layer norm"


We would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Before this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        14.10%           449.257ms        14.10%           449.257ms        9.360ms          NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```


After this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        1.55%            42.453ms         1.55%            42.453ms         884.448us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]     
```

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][aten] Enable the intra-op parallelism for layer norm"


We would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Before this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        14.10%           449.257ms        14.10%           449.257ms        9.360ms          NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]
```


After this Diff:
```
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
native_layer_norm        1.55%            42.453ms         1.55%            42.453ms         884.448us        NaN              0.000us          0.000us          48               [[61, 64, 1024], [1024], [1024]]     
```

Differential Revision: [D18063407](https://our.internmc.facebook.com/intern/diff/D18063407/)

[ghstack-poisoned]

xiaomengy reviewed (approved):
LGTM
